### PR TITLE
fix: AutoRAG context correctness always 0 for .txt input documents

### PR DESCRIPTION
**Pull request by witold-nowogorski**
Related issue: https://redhat.atlassian.net/browse/RHOAIENG-60783

**Description of your changes:**

Two bugs caused `context_correctness` to score 0 when using .txt input documents:

  1. text_extraction appended .md to every output file, so `hash_0.txt` became `hash_0.txt.md`. ai4rag then added another .md internally, producing `hash_0.txt.md.md` as the document ID - mismatching the benchmark's `hash_0.txt`.
  2. documents_discovery sampling priority compared full S3 keys (e.g. `docs/hash_0.txt`) against bare benchmark filenames (`hash_0.txt`), so priority sorting never worked.

  Changes

  - text_extraction: .txt files now keep their original extension; other formats still produce <name>.md.
  - documents_indexing / search_space_preparation / rag_templates_optimization: document ID derivation updated to match - stem for .md files, name for .txt files.
  - documents_discovery: sampling sort now uses `Path(c["Key"]).name` to match against benchmark document IDs.
  - Tests: existing tests updated; 12 new sampling tests added.

**Checklist:**

### Pre-Submission Checklist

- [ ] All tests and CI checks pass
- [ ] Pre-commit hooks pass without errors
- [ ] You have [signed off your commits](https://www.kubeflow.org/docs/about/contributing/#sign-off-your-commits)
- [ ] The title for your pull request (PR) should follow our title convention.
  [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### Additional Checklist Items for New or Updated Components/Pipelines

- [ ] `metadata.yaml` includes fresh `lastVerified` timestamp
- [ ] All [required files](https://github.com/kubeflow/pipelines-components/blob/main/docs/CONTRIBUTING.md#required-files)
  are present and complete
- [ ] OWNERS file lists appropriate maintainers
- [ ] README provides clear documentation with usage examples
- [ ] Component follows `snake_case` naming convention
- [ ] No security vulnerabilities in dependencies
- [ ] Containerfile included if using a custom base image

## Summary by CodeRabbit

* **New Features**
  * Preserve plain-text (.txt) files in the document pipeline and discover Markdown files case‑insensitively.

* **Bug Fixes**
  * Corrected document identifier handling so non‑Markdown filenames keep their extensions.
  * Document sampling now respects provided test-data ordering when prioritizing documents.

* **Tests**
  * Expanded tests for sampling behavior, size caps, and test-data prioritization.

### CODE REVIEW

**Review by DorotaDR — COMMENTED** [review]
@witold-nowogorski Please respond to the comments from CodeRabbit; after that, I can approve the PR.

**Comment by DorotaDR** [comment]
Please also share the results of the integration / e2e AutoRAG tests showing that the context correctness is non-zero.

File: components/data_processing/autorag/documents_discovery/component.py [hunk 1]
```diff
@@ -105,7 +106,7 @@ def _make_s3_client(verify=True):
 
     test_data_docs_names = get_test_data_docs_names()
     if test_data_docs_names:
-        supported_files.sort(key=lambda c: c["Key"] not in test_data_docs_names)
+        supported_files.sort(key=lambda c: Path(c["Key"]).name not in test_data_docs_names)
```
Comment (filip-komarzyniec) [inline components/data_processing/autorag/documents_discovery/component.py]: I guess what CodeRabbit points out here is that the logic might be incorrect when there are files with the same base names (which in some cases will be valid in S3 storage). 

How do we ensure we correctly process same-named files? This might be worth addressing somehow. 

The example can be the following folder structure:

```
documents/
├─ sub_folder1/
│  ├─ document1.txt   # this one is included in the benchmark.json file
├─ sub_folder2/
│  ├─ document1.txt  # this one is NOT included in the benchmark.json file

```

File: components/data_processing/autorag/documents_indexing/component.py [hunk 2]
```diff
@@ -129,7 +129,7 @@ def _create_llama_stack_client(**kwargs) -> LlamaStackClient:
         api_key=os.getenv("LLAMA_STACK_CLIENT_API_KEY"),
     )
 
-    paths = sorted(Path(extracted_text.path).glob("*.md"))
+    paths = sorted(p for ext in ("*.md", "*.txt") for p in Path(extracted_text.path).glob(ext))
```
Comment (filip-komarzyniec) [inline components/data_processing/autorag/documents_indexing/component.py]: Either this is not accurate or docstrings are out-of-date. In the `extracted_text` folder do we store only `*md` files or `*.txt` as well?

File: components/data_processing/autorag/documents_indexing/component.py [hunk 3]
```diff
@@ -157,7 +157,7 @@ def _create_llama_stack_client(**kwargs) -> LlamaStackClient:
         batch_documents = [
             Document(
                 page_content=p.read_text(encoding="utf-8", errors="replace"),
-                metadata={"document_id": p.stem},
+                metadata={"document_id": p.stem if p.suffix == ".md" else p.name},
```
Comment (filip-komarzyniec) [inline components/data_processing/autorag/documents_indexing/component.py]: Why exactly do we treat `*.md` files differently than `*.txt` ones? 

If this is because some other component appends or not extensions then I'd say we change the other component so that it correctly decided whether to append extension or not. 

The current approach is complicated and creates weird coupling dependant on the other component's internal logic.

File: components/data_processing/autorag/text_extraction/component.py [hunk 4]
```diff
@@ -234,12 +234,13 @@ def worker_process_document(file_path_str: str, output_dir_str: str) -> tuple[bo
...
             input_file = Path(file_path_str)
             output_dir = Path(output_dir_str)
-            output_file = output_dir / f"{input_file.name}.md"
-
             if input_file.suffix.lower() == ".txt":
+                output_file = output_dir / input_file.name
                 output_file.write_text(input_file.read_text(encoding="utf-8"), encoding="utf-8")
                 return True, None
 
+            output_file = output_dir / f"{input_file.name}.md"
```
Comment (filip-komarzyniec) [inline components/data_processing/autorag/text_extraction/component.py]: Is there any reason we keep differentiating between `*.txt` and `*.md` files? 

Wouldn't it be easier to write everything as `*.md`? What we wanted to achieve here is not to docling-process `*.txt` files as it makes no sense (and afaik is not really possible).

**Review by filip-komarzyniec — COMMENTED** [review]
The PR seems to be introducing many `if` clauses differentiating between `txt` and `md` files extensions. 
I'm not really sure this is the most optimal approach. I've left some comments to think through -- maybe we can make it work without so many `if`-s as they complicate the overall logic and open doors for more bugs. 

Please also link a successful test run (or any other sign that the proposed changes were tested).

File: components/data_processing/autorag/documents_indexing/component.py (hunk 2 above)
Comment (witold-nowogorski) [reply components/data_processing/autorag/documents_indexing/component.py]: `text_extraction` writes all outputs as *.md (.txt files are copied as-is but saved with a .md extension). The docstring has been updated to reflect this.

File: components/data_processing/autorag/documents_indexing/component.py (hunk 3 above)
Comment (witold-nowogorski) [reply components/data_processing/autorag/documents_indexing/component.py]: The root cause was text_extraction preserving the .txt extension for plain-text files while appending .md for everything else. I changed it to write all outputs uniformly as <original_name>.md, which removed the need for any conditional in documents_indexing.

File: components/data_processing/autorag/text_extraction/tests/test_component_unit.py [hunk 5]
```diff
@@ -474,6 +474,7 @@ def fake_download_file(bucket, key, dest):
         output_dir = Path(extracted_text_artifact.path)
         output_files = list(output_dir.glob("*.md"))
         assert len(output_files) == 1
+        assert output_files[0].name == "file1.txt.md"
```
[the remark this answers is not in the document]
Comment (witold-nowogorski) [reply components/data_processing/autorag/text_extraction/tests/test_component_unit.py:477]: This finding is not valid for the current state of the PR. The tests were intentionally updated to expect `*.txt.md` output as part of this change - text_extraction now writes all files uniformly as `<original_name>.md`, including plain-text files. This was agreed upon in the review to remove the .txt vs .md branching. Reverting the assertions to expect file1.txt would lock in the old behavior that was deliberately changed.

**Comment by LukaszCmielowski** [comment]
@filip-komarzyniec please finalize the review.

File: components/data_processing/autorag/documents_discovery/component.py [hunk 6]
```diff
@@ -105,7 +106,8 @@ def _make_s3_client(verify=True):
 
     test_data_docs_names = get_test_data_docs_names()
     if test_data_docs_names:
-        supported_files.sort(key=lambda c: c["Key"] not in test_data_docs_names)
+        test_keys_set = {c["Key"] for c in supported_files if Path(c["Key"]).name in test_data_docs_names}
```
Comment (filip-komarzyniec) [inline components/data_processing/autorag/documents_discovery/component.py:109]: I guess what CodeRabbit pointed out here is that the logic might be incorrect when there are files with the same base names (which in some cases will be valid in S3 storage). 

How do we ensure we correctly process same-named files? This might be worth addressing somehow. 

The example can be the following folder structure:

```
documents/
├─ sub_folder1/
│  ├─ document1.txt
├─ sub_folder2/
│  ├─ document1.txt

```

File: components/data_processing/autorag/documents_indexing/component.py [hunk 7]
```diff
@@ -129,7 +129,8 @@ def _create_llama_stack_client(**kwargs) -> LlamaStackClient:
         api_key=os.getenv("LLAMA_STACK_CLIENT_API_KEY"),
     )
 
-    paths = sorted(Path(extracted_text.path).glob("*.md"))
+    base = Path(extracted_text.path)
+    paths = sorted(p for p in base.iterdir() if p.is_file() and p.suffix.lower() == ".md")
```
Comment (filip-komarzyniec) [inline components/data_processing/autorag/documents_indexing/component.py:133]: If `extracted_text` directory stores only `*.md` files now  then why do we need `and p.suffix.lower() == ".md"` check?

File: components/training/autorag/rag_templates_optimization/component.py [hunk 8]
```diff
@@ -593,16 +593,18 @@ def load_as_langchain_doc(path: str | Path) -> list[Document]:
         if path.is_dir():
             for doc_path in path.iterdir():
                 with doc_path.open("r", encoding="utf-8") as doc:
+                    doc_id = doc_path.stem if doc_path.suffix == ".md" else doc_path.name
```
Comment (filip-komarzyniec) [inline components/training/autorag/rag_templates_optimization/component.py:596]: Again, why do we need the `*.md` checks now? 

According to discussion in some previous comments all files in `extracted_text` should be markdown

**Comment by filip-komarzyniec** [comment]
a couple more comments but nothing really severe. Apart from that the changes /lgtm

**Comment by LukaszCmielowski** [comment]
@witold-nowogorski can we have those addressed so I could merge ?.

File: components/data_processing/autorag/documents_indexing/component.py (hunk 7 above)
Comment (witold-nowogorski) [reply components/data_processing/autorag/documents_indexing/component.py:133]: It might be redundant but I will preserve it just to make sure we are loading markdowns only in case something else would appear in the `extracted_text` directory, such as some OS artifacts, some metadata files etc.

File: components/training/autorag/rag_templates_optimization/component.py (hunk 8 above)
Comment (witold-nowogorski) [reply components/training/autorag/rag_templates_optimization/component.py:596]: Agreed, it's redundant today, but cheap enough to keep as a guard against unexpected files appearing in `extracted_text` (OS artifacts, future changes or something dropped silently by text extraction process).

File: components/data_processing/autorag/documents_indexing/component.py (hunk 7 above)
Comment (filip-komarzyniec) [reply components/data_processing/autorag/documents_indexing/component.py:133]: makes sense!
